compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -66,6 +66,9 @@ object Typer {
     if (!tree.isEmpty && !tree.isInstanceOf[untpd.TypedSplice] && ctx.typerState.isGlobalCommittable)
       assert(tree.span.exists, i"position not set for $tree # ${tree.uniqueId} of ${tree.getClass} in ${tree.source}")
 
+  /** An attachment for GADT constraints that were inferred for a pattern. */
+  val InferredGadtConstraints = new Property.Key[core.GadtConstraint]
```

Comment:
This should be a StickyKey to not get lost through tree transformations:
```suggestion
  val InferredGadtConstraints = new Property.StickyKey[core.GadtConstraint]
```